src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs

```diff
@@ -1,48 +1,129 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Collections.Generic;
-using System.Linq;
+using System.Reflection;
 using Xunit;
 
-namespace System.IO.Compression.Tests
+namespace System.IO.Compression.Tests;
+
+[Collection(nameof(DisableParallelization))]
+public class zip_LargeFiles : ZipFileTestBase
 {
-    [Collection(nameof(DisableParallelization))]
-    public class zip_LargeFiles : ZipFileTestBase
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsSpeedOptimized), nameof(PlatformDetection.Is64BitProcess))] // don't run it on slower runtimes
+    [OuterLoop("It requires almost 12 GB of free disk space")]
+    public static void UnzipOver4GBZipFile()
     {
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsSpeedOptimized), nameof(PlatformDetection.Is64BitProcess))] // don't run it on slower runtimes
-        [OuterLoop("It requires almost 12 GB of free disk space")]
-        public static void UnzipOver4GBZipFile()
+        byte[] buffer = GC.AllocateUninitializedArray<byte>(1_000_000_000); // 1 GB
+
+        string zipArchivePath = Path.Combine(Path.GetTempPath(), "over4GB.zip");
+        DirectoryInfo tempDir = Directory.CreateDirectory(Path.Combine(Path.GetTempPath(), "over4GB"));
+
+        try
+        {
+            for (byte i = 0; i < 6; i++)
+            {
+                File.WriteAllBytes(Path.Combine(tempDir.FullName, $"{i}.test"), buffer);
+            }
+
+            ZipFile.CreateFromDirectory(tempDir.FullName, zipArchivePath, CompressionLevel.NoCompression, includeBaseDirectory: false);
+
+            using ZipArchive zipArchive = ZipFile.OpenRead(zipArchivePath);
+            foreach (ZipArchiveEntry entry in zipArchive.Entries)
+            {
+                using Stream entryStream = entry.Open();
+
+                Assert.True(entryStream.CanRead);
+                Assert.Equal(buffer.Length, entryStream.Length);
+            }
+        }
+        finally
         {
-            byte[] buffer = GC.AllocateUninitializedArray<byte>(1_000_000_000); // 1 GB
+            File.Delete(zipArchivePath);
+
+            tempDir.Delete(recursive: true);
+        }
+    }
+
+    private static void FillWithHardToCompressData(byte[] buffer)
+    {
+        Random.Shared.NextBytes(buffer);
+    }
 
-            string zipArchivePath = Path.Combine(Path.GetTempPath(), "over4GB.zip");
-            DirectoryInfo tempDir = Directory.CreateDirectory(Path.Combine(Path.GetTempPath(), "over4GB"));
+    [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsSpeedOptimized), nameof(PlatformDetection.Is64BitProcess))] // don't run it on slower runtimes
+    [OuterLoop("It requires 5~6 GB of free disk space and a lot of CPU time for compressed tests")]
+    [InlineData(false)]
+    [InlineData(true)]
+    public static void CheckZIP64VersionIsSet_ForSmallFilesAfterBigFiles(bool isCompressed)
```

Comment:
Can we also make sure we continue to open the "bad" zips that were produced prior to this fix?  I think we should, but it would be good to capture that in a regression test.

Reply:
Sure. I'm currently writing this test locally. Will submit it as a separate PR in main, and will then backport it, so that I don't block this.

Reply:
Confirmed in main in this PR: https://github.com/dotnet/runtime/pull/103152